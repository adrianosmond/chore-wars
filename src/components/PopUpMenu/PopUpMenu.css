.pop-up-menu {
  
}

.pop-up-menu__button {
  padding: 0;
  margin: 0;
  border: 0;
  outline: 0;
  background-color: transparent;
  cursor: pointer;
  appearance: none;
}

.pop-up-menu__menu {
  position: absolute;
  z-index: 2;
  top: 2.5rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  background-color: #fff;
  box-shadow: 0 0 10px #666;
  transform: scale(0);
  transition: transform 0.1s;
}

.pop-up-menu__menu--left {
  left: 0.5rem;
  transform-origin: 0 0;
}

.pop-up-menu__menu--right {
  right: 0.5rem;
  transform-origin: 100% 0;
}

.pop-up-menu__menu--visible {
  transform: scale(1);
}

.pop-up-menu__menu::before {
  content: "";
  position: absolute;
  top: -0.5rem;
  right: 0.25rem;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 0 0.5rem 0.5rem 0.5rem;
  border-color: transparent transparent #fff transparent;
}

.pop-up-menu__menu--left::before {
  left: 0.25rem;
}

.pop-up-menu__menu--right::before {
  right: 0.25rem;
}

.pop-up-menu__item {
  display: block;
  width: 100%;
  padding: 0.5rem 1rem;
  font: inherit;
  border: 0;
  background-color: transparent;
  text-decoration: none;
  color: #000;
  white-space: nowrap;
  text-align: right;
  cursor: pointer;
  -webkit-appearance: none;
}

.pop-up-menu__item:hover,
.pop-up-menu__item:focus {
  background-color: #f7f7f7;
}
