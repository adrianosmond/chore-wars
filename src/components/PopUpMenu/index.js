import React, { Component } from 'react';
import { Link } from 'react-router-dom';

import './index.css';

const makeLink = (child, idx) =>
  <Link to={child.to} className="pop-up-menu__item" key={idx}>{child.text}</Link>;

const makeButton = (child, idx, onClick) =>
  <button onClick={onClick} className="pop-up-menu__item" key={idx}>{child.text}</button>;

class PopUpMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      side: props.side || 'right',
      visible: false,
    };
  }

  toggleMenu() {
    this.setState({
      visible: !this.state.visible,
    });
  }

  render() {
    const { side, visible } = this.state;
    return (
      <div className={`pop-up-menu ${this.props.extraClasses}`}>
        <button className="pop-up-menu__button" onClick={this.toggleMenu.bind(this)} data-test="toggle-menu">
          { this.props.children }
        </button>
        <div className={`pop-up-menu__menu pop-up-menu__menu--${side}${visible ? ' pop-up-menu__menu--visible' : ''}`}>
          { this.props.options.map((child, idx) => {
            if (child.type === 'link') {
              return makeLink(child, idx);
            } else if (child.type === 'button') {
              return makeButton(child, idx, () => {
                this.toggleMenu();
                child.onClick();
              });
            }
            return null;
          }) }
        </div>
      </div>
    );
  }
}

export default PopUpMenu;
