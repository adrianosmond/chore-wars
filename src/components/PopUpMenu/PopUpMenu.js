import React, { Component } from 'react';
import { Link } from 'react-router-dom';

import './PopUpMenu.css';

const makeLink = (child, idx) => (
  <Link to={child.to} className="pop-up-menu__item" key={idx}>{child.text}</Link>
);

const makeButton = (child, idx, onClick) => (
  <button type="button" onClick={onClick} className="pop-up-menu__item" key={idx}>{child.text}</button>
);

class PopUpMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      side: props.side || 'right',
      visible: false,
    };
  }

  toggleMenu() {
    const { visible } = this.state;
    this.setState({
      visible: !visible,
    });
  }

  render() {
    const { side, visible } = this.state;
    const { extraClasses, children, options } = this.props;
    return (
      <div className={`pop-up-menu ${extraClasses}`}>
        <button
          type="button"
          className="pop-up-menu__button"
          onClick={this.toggleMenu.bind(this)}
          data-test="toggle-menu"
        >
          { children }
        </button>
        <div className={`pop-up-menu__menu pop-up-menu__menu--${side}${visible ? ' pop-up-menu__menu--visible' : ''}`}>
          { options.map((child, idx) => {
            if (child.type === 'link') {
              return makeLink(child, idx);
            } if (child.type === 'button') {
              return makeButton(child, idx, () => {
                this.toggleMenu();
                child.onClick();
              });
            }
            return null;
          }) }
        </div>
      </div>
    );
  }
}

export default PopUpMenu;
