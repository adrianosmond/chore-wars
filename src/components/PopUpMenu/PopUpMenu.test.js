import React from 'react';
import { shallow } from 'enzyme';
import PopUpMenu from './index';

describe('Popup Menu', () => {
  const mockButtonClick = jest.fn();
  const popUpMenu = shallow(<PopUpMenu options={[
      {
        type: 'link',
        to: '/link/address',
        text: 'link',
      },
      {
        type: 'button',
        onClick: mockButtonClick,
        text: 'link',
      },
    ]}>TRIGGER</PopUpMenu>);
  const toggleButton = popUpMenu.find('[data-test="toggle-menu"]');

  it('Renders a new chore menu', () => {
    expect(popUpMenu).toMatchSnapshot();
  });

  it('Can be toggled', () => {
    toggleButton.simulate('click');
    expect(popUpMenu.state().visible).toBe(true);
    expect(popUpMenu).toMatchSnapshot();
    toggleButton.simulate('click');
    expect(popUpMenu.state().visible).toBe(false);
  });

  it('Can have clickable buttons which close the menu', () => {
    toggleButton.simulate('click');
    expect(popUpMenu.state().visible).toBe(true);
    const button = popUpMenu.find('button.pop-up-menu__item');
    button.simulate('click');
    expect(mockButtonClick).toHaveBeenCalled();
    expect(popUpMenu.state().visible).toBe(false);
  });
});
