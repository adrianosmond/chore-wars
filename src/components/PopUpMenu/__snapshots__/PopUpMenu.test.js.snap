// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Menu Can be toggled 1`] = `
<div
  className="pop-up-menu undefined"
>
  <button
    className="pop-up-menu__button"
    data-test="toggle-menu"
    onClick={[Function]}
  >
    TRIGGER
  </button>
  <div
    className="pop-up-menu__menu pop-up-menu__menu--right pop-up-menu__menu--visible"
  >
    <Link
      className="pop-up-menu__item"
      key="0"
      replace={false}
      to="/link/address"
    >
      link
    </Link>
    <button
      className="pop-up-menu__item"
      key="1"
      onClick={[Function]}
    >
      link
    </button>
  </div>
</div>
`;

exports[`Popup Menu Renders a new chore menu 1`] = `
<div
  className="pop-up-menu undefined"
>
  <button
    className="pop-up-menu__button"
    data-test="toggle-menu"
    onClick={[Function]}
  >
    TRIGGER
  </button>
  <div
    className="pop-up-menu__menu pop-up-menu__menu--right"
  >
    <Link
      className="pop-up-menu__item"
      key="0"
      replace={false}
      to="/link/address"
    >
      link
    </Link>
    <button
      className="pop-up-menu__item"
      key="1"
      onClick={[Function]}
    >
      link
    </button>
  </div>
</div>
`;
